File: setting.py
import json
from os import getcwd

from define import Playmodes

setting_filepath = 'setting.json'

default = {
    'newrecord_only': False,
    'play_sound': False,
    'autosave': False,
    'autosave_filtered': False,
    'filter_compact': False,
    'filter_overlay': {
        'use': False,
        'rival': {
            'imagefilepath': None,
            'offset': (0, 0, ),
            'scalefactor': 1,
        },
        'loveletter': {
            'imagefilepath': None,
            'offset': (0, 0, ),
            'scalefactor': 1,
        },
        'rivalname': {
            'imagefilepath': None,
            'offset': (0, 0, ),
            'scalefactor': 1,
        },
    },
    'savefilemusicname_right': False,
    'hotkeys': {
        'active_screenshot': 'alt+F10',
        'select_summary': 'alt+U',
        'select_notesradar': 'alt+R',
        'select_screenshot': 'alt+T',
        'select_scoreinformation': 'alt+I',
        'select_scoregraph': 'alt+G',
        'upload_musicselect': 'alt+F8',
    },
    'summary_countmethod_only': False,
    'display_result': False,
    'resultimage_filtered': False,
    'imagesave_path': getcwd(),
    'startup_image': 'notesradar',
    'hashtags': '#IIDX #infinitas573 #infnotebook',
    'data_collection': False,
    'summaries' : {
        Playmodes.SP: {
            '8': {'cleartypes': ['F-COMBO'], 'djlevels': ['AAA']},
            '9': {'cleartypes': ['F-COMBO'], 'djlevels': ['AAA']},
            '10': {'cleartypes': ['F-COMBO'], 'djlevels': ['AAA']},
            '11': {'cleartypes': ['F-COMBO'], 'djlevels': ['AAA']},
            '12': {'cleartypes': ['F-COMBO'], 'djlevels': ['AAA']},
        },
        Playmodes.DP: {
            '8': {'cleartypes': ['F-COMBO'], 'djlevels': ['AAA']},
            '9': {'cleartypes': ['F-COMBO'], 'djlevels': ['AAA']},
            '10': {'cleartypes': ['F-COMBO'], 'djlevels': ['AAA']},
            '11': {'cleartypes': ['F-COMBO'], 'djlevels': ['AAA']},
            '12': {'cleartypes': ['F-COMBO'], 'djlevels': ['AAA']},
        },
    },
    'discord_webhook': {
        'playername': 'NO NAME',
        'seenevents': [],
        'joinedevents': {},
    },
    'port': {
        'main': 52374,
        'socket': 57328,
    },
}

class Setting():
    def __init__(self):
        try:
            with open(setting_filepath) as f:
                self.json = json.load(f)
        except Exception:
            self.json = default

        for k in default.keys():
            if not k in self.json.keys():
                self.json[k] = default[k]
        
        # バージョンアップによる仕様変更の対処
        if 'display_saved_result' in self.json.keys():
            self.json['display_result'] = self.json['display_saved_result']
            del self.json['display_saved_result']
        if 'save_newrecord_only' in self.json.keys():
            self.json['newrecord_only'] = self.json['save_newrecord_only']
            del self.json['save_newrecord_only']
        if 'autoexport' in self.json.keys():
            del self.json['autoexport']
        if 'manage' in self.json.keys():
            self.json['debug'] = self.json['manage']
            del self.json['manage']
        if 'display_music' in self.json.keys():
            del self.json['display_music']
        if 'djname' in self.json['discord_webhook'].keys():
            self.json['discord_webhook']['playername'] = self.json['discord_webhook']['djname']
            del self.json['discord_webhook']['djname']
        if 'servers' in self.json['discord_webhook'].keys():
            del self.json['discord_webhook']['servers']
        if not 'playername' in self.json['discord_webhook'].keys():
            self.json['discord_webhook']['playername'] = default['discord_webhook']['playername']
        if not 'seenevents' in self.json['discord_webhook'].keys():
            self.json['discord_webhook']['seenevents'] = []
        if not 'joinedevents' in self.json['discord_webhook'].keys():
            self.json['discord_webhook']['joinedevents'] = {}
        if not 'select_summary' in self.json['hotkeys'].keys():
            self.json['hotkeys']['select_summary'] = default['hotkeys']['select_summary']
        if not 'select_notesradar' in self.json['hotkeys'].keys():
            self.json['hotkeys']['select_notesradar'] = default['hotkeys']['select_notesradar']
        if not 'select_screenshot' in self.json['hotkeys'].keys():
            self.json['hotkeys']['select_screenshot'] = default['hotkeys']['select_screenshot']
        if not 'select_scoreinformation' in self.json['hotkeys'].keys():
            self.json['hotkeys']['select_scoreinformation'] = default['hotkeys']['select_scoreinformation']
        if not 'select_scoregraph' in self.json['hotkeys'].keys():
            self.json['hotkeys']['select_scoregraph'] = default['hotkeys']['select_scoregraph']
        if not 'filter_overlay' in self.json.keys():
            self.json['filter_overlay'] = default['filter_overlay']
        if not 'use' in self.json['filter_overlay'].keys():
            self.json['filter_overlay']['use'] = default['filter_overlay']['use']
        for key in ['rival', 'loveletter', 'rivalname']:
            if not key in self.json['filter_overlay'].keys():
                self.json['filter_overlay'][key] = default['filter_overlay'][key]
            if not 'imagefilepath' in self.json['filter_overlay'][key].keys():
                self.json['filter_overlay'][key]['imagefilepath'] = default['filter_overlay'][key]['imagefilepath']
            if not 'offset' in self.json['filter_overlay'][key].keys():
                self.json['filter_overlay'][key]['offset'] = default['filter_overlay'][key]['offset']
            if not 'scalefactor' in self.json['filter_overlay'][key].keys():
                self.json['filter_overlay'][key]['scalefactor'] = default['filter_overlay'][key]['scalefactor']

        self.save()

    def save(self):
        try:
            with open(setting_filepath, 'w') as f:
                json.dump(self.json, f, indent=2)
        except Exception as ex:
            print(f'setting json dump error: {ex}')
    
    def has_key(self, key):
        return key in self.json.keys()

    def get_value(self, key):
        if not key in self.json.keys():
            return False
        
        return self.json[key]

    def set_value(self, key, value):
        self.json[key] = value

    @property
    def newrecord_only(self):
        '''更新があったときのみ記録する
        '''
        return self.get_value('newrecord_only')

    @newrecord_only.setter
    def newrecord_only(self, value: bool):
        self.set_value('newrecord_only', value)
    
    @property
    def play_sound(self):
        '''リザルトを認識した時等に音を鳴らす
        '''
        return self.get_value('play_sound')

    @play_sound.setter
    def play_sound(self, value: bool):
        self.set_value('play_sound', value)
    
    @property
    def autosave(self):
        '''リザルト画像を自動保存する
        '''
        return self.get_value('autosave')

    @autosave.setter
    def autosave(self, value: bool):
        self.set_value('autosave', value)
    
    @property
    def autosave_filtered(self):
        '''ライバルを隠したリザルト画像を自動保存する
        '''
        return self.get_value('autosave_filtered')

    @autosave_filtered.setter
    def autosave_filtered(self, value: bool):
        self.set_value('autosave_filtered', value)
    
    @property
    def filter_compact(self):
        '''ライバル隠しを最小にする
        '''
        return self.get_value('filter_compact')

    @filter_compact.setter
    def filter_compact(self, value: bool):
        self.set_value('filter_compact', value)
    
    @property
    def filter_overlay(self):
        '''ライバル隠しに重ねる画像の設定
        '''
        return self.get_value('filter_overlay')

    @filter_overlay.setter
    def filter_overlay(self, value: dict[str, any]):
        self.set_value('filter_overlay', value)
    
    @property
    def savefilemusicname_right(self):
        '''画像ファイルのファイル名につける曲名を最後にする
        '''
        return self.get_value('savefilemusicname_right')

    @savefilemusicname_right.setter
    def savefilemusicname_right(self, value: bool):
        self.set_value('savefilemusicname_right', value)
    
    @property
    def hotkeys(self):
        '''ショートカットキーの設定
        '''
        return self.get_value('hotkeys')

    @hotkeys.setter
    def hotkeys(self, value: dict[str, str]):
        self.set_value('hotkeys', value)
    
    @property
    def summary_countmethod_only(self):
        '''統計の譜面数のカウント方式が、該当するもののみ

        そうでない場合は、例えばEXH-CLEAR等もF-COMBOに含める。
        '''
        return self.get_value('summary_countmethod_only')

    @summary_countmethod_only.setter
    def summary_countmethod_only(self, value: bool):
        self.set_value('summary_countmethod_only', value)
    
    @property
    def display_result(self):
        '''リザルトを記録時にそのリザルトを選択する
        '''
        return self.get_value('display_result')

    @display_result.setter
    def display_result(self, value: bool):
        self.set_value('display_result', value)
    
    @property
    def resultimage_filtered(self):
        '''表示リザルト画像のフィルター
        '''
        return self.get_value('resultimage_filtered')

    @resultimage_filtered.setter
    def resultimage_filtered(self, value):
        self.set_value('resultimage_filtered', value)

    @property
    def imagesave_path(self):
        '''リザルト画像等のファイルの保存先のフォルダパス
        '''
        return self.get_value('imagesave_path')

    @imagesave_path.setter
    def imagesave_path(self, value: str):
        self.set_value('imagesave_path', value)
    
    @property
    def startup_image(self):
        '''リザルト手帳の初期処理完了後に選択状態にするタブの設定
        '''
        return self.get_value('startup_image')
    
    @startup_image.setter
    def startup_image(self, value: str):
        self.set_value('startup_image', value)
    
    @property
    def hashtags(self):
        '''ポストの本文に付与する文字列
        '''
        return self.get_value('hashtags')
    
    @hashtags.setter
    def hashtags(self, value: str):
        self.set_value('hashtags', value)
    
    @property
    def data_collection(self):
        '''収集画像をアップロードする
        '''
        return self.get_value('data_collection')

    @data_collection.setter
    def data_collection(self, value: bool):
        self.set_value('data_collection', value)
    
    @property
    def summaries(self):
        '''統計出力設定
        '''
        return self.get_value('summaries')

    @summaries.setter
    def summaries(self, value):
        self.set_value('summaries', value)
    
    @property
    def discord_webhook(self):
        '''イベント設定
        '''
        return self.get_value('discord_webhook')

    @discord_webhook.setter
    def discord_webhook(self, value):
        self.set_value('discord_webhook', value)

    @property
    def port(self):
        '''ポート設定
        '''
        return self.get_value('port')

    @port.setter
    def port(self, value):
        self.set_value('port', value)

    @property
    def ignore_download(self):
        '''最新リソースファイルの取得を無効にする
        '''
        return self.get_value('ignore_download')

    @property
    def debug(self):
        '''デバッグモード

        デバッグ用のログ表示を有効にする。
        '''
        return self.get_value('debug')

File: web/setting.js
let setting = null;

$(function() {
  webui.setEventCallback((e) => {
    if(e == webui.event.CONNECTED) initialize();
    if(e == webui.event.DISCONNECTED) console.log('disconnect.');
  });

  $('button.select_tabpage').on('click', onclick_tab);

  $('#check_filter_overlay').on('change', onchange_check_filteroverlay);
  $('button.button_browse_file').on('click', onclick_browse_file);
  $('button.button_browse_directory').on('click', onclick_browse_directory);
  
  $('button#button_save').on('click', onclick_button_save);
  $('button#button_close').on('click', onclick_button_close);

  switch_displaytab('blank');
});

/**
 * 初期処理
 * 
 * ロード完了時に実行する。
 */
async function initialize() {
  document.body.addEventListener('contextmenu', e => e.stopPropagation(), true);

  const setting = JSON.parse(await webui.get_setting());
  globalThis.setting = setting;
  
  $('#check_newrecord_only').prop('checked', setting['newrecord_only']);
  $('#check_play_sound').prop('checked', setting['play_sound']);
  $('#check_autosave').prop('checked', setting['autosave']);
  $('#check_autosave_filtered').prop('checked', setting['autosave_filtered']);
  $('#check_filter_compact').prop('checked', setting['filter_compact']);
  $('#check_filter_overlay').prop('checked', setting['filter_overlay']['use']);
  ['rival', 'loveletter', 'rivalname'].forEach(key => {
    $(`#text_overlayimage${key}_path`).val(setting['filter_overlay'][key]['imagefilepath']);
    $(`#text_overlayimage${key}_offsetx`).val(setting['filter_overlay'][key]['offset'][0]);
    $(`#text_overlayimage${key}_offsety`).val(setting['filter_overlay'][key]['offset'][1]);
    $(`#text_overlayimage${key}_scalefactor`).val(setting['filter_overlay'][key]['scalefactor']);
  });

  $('#check_savefilemusicname_right').prop('checked', setting['savefilemusicname_right']);

  $('#text_hotkey_active_screenshot').val(setting['hotkeys']['active_screenshot']);
  $('#text_hotkey_select_summary').val(setting['hotkeys']['select_summary']);
  $('#text_hotkey_select_notesradar').val(setting['hotkeys']['select_notesradar']);
  $('#text_hotkey_select_screenshot').val(setting['hotkeys']['select_screenshot']);
  $('#text_hotkey_select_scoreinformation').val(setting['hotkeys']['select_scoreinformation']);
  $('#text_hotkey_select_scoregraph').val(setting['hotkeys']['select_scoregraph']);
  $('#text_hotkey_upload_musicselect').val(setting['hotkeys']['upload_musicselect']);

  if(!setting['summary_countmethod_only'])
    $('#radio_summary_countmethod_sum').prop('checked', true);
  else
    $('#radio_summary_countmethod_only').prop('checked', true);

  $('#check_display_result').prop('checked', setting['display_result']);
  $('#check_resultimage_filtered').prop('checked', setting['resultimage_filtered']);

  $('#text_imagesave_path').val(setting['imagesave_path']);

  $('#radio_startup_image_notesradar').prop('checked', setting['startup_image'] == 'notesradar');
  $('#radio_startup_image_summary').prop('checked', setting['startup_image'] == 'summary');

  $('#text_hashtags').val(setting['hashtags']);
  $('#check_data_collection').prop('checked', setting['data_collection']);

  $('input#text_portmain').val(setting['port']['main']);
  $('input#text_portsocket').val(setting['port']['socket']);

  $('div.summary_group_playmode').each(function() {
    const playmode = $(this).children('div.summary_playmode').text();
    if(!setting['summaries'].hasOwnProperty(playmode)) return true;

    $(this).find('div.summary_levelblock div.summary_group_level').each(function() {
      const level = $(this).children('div.summary_level').text();
      if(!setting['summaries'][playmode].hasOwnProperty(level)) return true;

      if(setting['summaries'][playmode][level].hasOwnProperty('cleartypes')) {
        $(this).find('div.summary_itemblock div.cleartype').each(function() {
          const key = $(this).children('label').text();
          if(setting['summaries'][playmode][level]['cleartypes'].includes(key))
            $(this).children('input').prop('checked', true);
        });
      }

      if(setting['summaries'][playmode][level].hasOwnProperty('djlevels')) {
        $(this).find('div.summary_itemblock div.djlevel').each(function() {
          const key = $(this).children('label').text();
          if(setting['summaries'][playmode][level]['djlevels'].includes(key))
            $(this).children('input').prop('checked', true);
        });
      }
    });
  });

  switch_filteroverlayenabled();
  
  const params = new URLSearchParams(window.location.search);
  if(!params.has('tab')) {
    switch_displaytab('general');
  }
  else {
    const tabname = params.get('tab')
    switch_displaytab(tabname);
  }
}

/**
 * タブの切り替え
 * 
 * 選択タブを指定のタブ名にする。
 * @param {str} tabname 対象のタブ名
 */
function switch_displaytab(tabname) {
  $('div.tabpage').css('display', 'none');
  $('button.select_tabpage').removeClass('select_tabpage_selected');

  $(`div#tabpage_${tabname}`).css('display', 'flex');
  $(`button#display_tabpage_${tabname}`).addClass('select_tabpage_selected');
}

/**
 * ライバル欄画像貼り付け設定の有効/無効の切替
 */
function switch_filteroverlayenabled(e) {
  const v = $('#check_filter_overlay').prop('checked');

  $('div.box_filter_overlay input').prop('disabled', !v);
  $('div.box_filter_overlay button').prop('disabled', !v);
}

/**
 * タブを選択
 * @param {ce.Event} e イベントハンドラ
 */
function onclick_tab(e) {
  const id = e.target.id;

  const extracted = id.match(/(?<=display_tabpage_)(.*)/)[0];
  const splitted = extracted.split('_');

  switch_displaytab(splitted[0]);
}

/**
 * ライバル欄画像貼り付け設定が変更された
 * @param {ce.Event} e イベントハンドラ
 */
function onchange_check_filteroverlay(e) {
  switch_filteroverlayenabled();
}

/**
 * ファイルを選択する
 * @param {ce.Event} e イベントハンドラ
 */
async function onclick_browse_file(e) {
  const targetelement = $(e.target).prev();
  const selected = JSON.parse(await webui.browse_file(
    targetelement.val().replaceAll('\\', '/'),
    JSON.stringify([['Image file', '*.png']]),
  ));

  if(selected == null) return;

  const normalized = selected.replaceAll('/', '\\');
  targetelement.val(normalized);
}

/**
 * フォルダを選択する
 * @param {ce.Event} e イベントハンドラ
 */
async function onclick_browse_directory(e) {
  const targetelement = $(e.target).prev();
  const selected = JSON.parse(await webui.browse_directory(
    targetelement.val().replaceAll('\\', '/'),
  ));

  if(selected == null) return;

  const normalized = selected.replaceAll('/', '\\');
  targetelement.val(normalized);
}

/**
 * 設定の保存かキャンセルボタンを押す
 * @param {ce.Event} e イベントハンドラ
 */
async function onclick_button_save(e) {
  const startupimage_selectedid = $('input[name="startup_image"]:checked').attr('id');
  const startup_image = startupimage_selectedid.match(/(?<=radio_startup_image_)(.*)/)[0];

  const summaries = {};
  $('div.summary_group_playmode').each(function() {
    const playmode = $(this).children('div.summary_playmode').text();

    summaries[playmode] = {}

    $(this).find('div.summary_levelblock div.summary_group_level').each(function() {
      const level = $(this).children('div.summary_level').text();

      summaries[playmode][level] = {
        'cleartypes': [],
        'djlevels': [],
      };

      $(this).find('div.summary_itemblock div.cleartype').each(function() {
        const key = $(this).children('label').text();
        if($(this).children('input').prop('checked'))
          summaries[playmode][level]['cleartypes'].push(key);
      });

      $(this).find('div.summary_itemblock div.djlevel').each(function() {
        const key = $(this).children('label').text();
        if($(this).children('input').prop('checked'))
          summaries[playmode][level]['djlevels'].push(key);
      });
    });
  });

  await webui.save_setting(JSON.stringify({
    'newrecord_only': $('#check_newrecord_only').prop('checked'),
    'play_sound': $('#check_play_sound').prop('checked'),
    'autosave': $('#check_autosave').prop('checked'),
    'autosave_filtered': $('#check_autosave_filtered').prop('checked'),
    'filter_compact': $('#check_filter_compact').prop('checked'),
    'filter_overlay': {
      'use': $('#check_filter_overlay').prop('checked'),
      'rival': {
        'imagefilepath': $('#text_overlayimagerival_path').val(),
        'offset': [
          $('#text_overlayimagerival_offsetx').val(),
          $('#text_overlayimagerival_offsety').val(),
        ],
        'scalefactor': $('#text_overlayimagerival_scalefactor').val(),
      },
      'loveletter': {
        'imagefilepath': $('#text_overlayimageloveletter_path').val(),
        'offset': [
          $('#text_overlayimageloveletter_offsetx').val(),
          $('#text_overlayimageloveletter_offsety').val(),
        ],
        'scalefactor': $('#text_overlayimageloveletter_scalefactor').val(),
      },
      'rivalname': {
        'imagefilepath': $('#text_overlayimagerivalname_path').val(),
        'offset': [
          $('#text_overlayimagerivalname_offsetx').val(),
          $('#text_overlayimagerivalname_offsety').val(),
        ],
        'scalefactor': $('#text_overlayimagerivalname_scalefactor').val(),
      },
    },
    'savefilemusicname_right': $('#check_savefilemusicname_right').prop('checked'),
    'hotkeys': {
      'active_screenshot': $('#text_hotkey_active_screenshot').val(),
      'select_summary': $('#text_hotkey_select_summary').val(),
      'select_notesradar': $('#text_hotkey_select_notesradar').val(),
      'select_screenshot': $('#text_hotkey_select_screenshot').val(),
      'select_scoreinformation': $('#text_hotkey_select_scoreinformation').val(),
      'select_scoregraph': $('#text_hotkey_select_scoregraph').val(),
      'upload_musicselect': $('#text_hotkey_upload_musicselect').val(),
    },
    'summary_countmethod_only': $('#radio_summary_countmethod_only').prop('checked'),
    'display_result': $('#check_display_result').prop('checked'),
    'resultimage_filtered': $('#check_resultimage_filtered').prop('checked'),
    'imagesave_path': $('#text_imagesave_path').val(),
    'startup_image': startup_image,
    'hashtags': $('#text_hashtags').val(),
    'data_collection': $('#check_data_collection').prop('checked'),
    'port': {
      'main': $('input#text_portmain').val(),
      'socket': $('input#text_portsocket').val(),
    },
    'summaries': summaries,
  }));

  window.parent.postMessage("setting_close", "*");
}

/**
 * 閉じるボタンを押す
 * 
 * ウィンドウを閉じる。
 * @param {ce.Event} e イベントハンドラ
 */
function onclick_button_close(e) {
  window.parent.postMessage("setting_close", "*");
}
